lib/IntermediateRepresentation/basicblock.h

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef FORTRAN_INTERMEDIATEREPRESENTATION_BASICBLOCK_H_
+#define FORTRAN_INTERMEDIATEREPRESENTATION_BASICBLOCK_H_
+
+#include "mixin.h"
+#include "region.h"
+#include <iostream>
+
+namespace Fortran::IntermediateRepresentation {
+
+struct Region;
+struct Statement;
+
+struct BasicBlock final : public llvm::ilist_node<BasicBlock>,
+                          public ChildMixin<BasicBlock, Region> {
+  using StatementListType = llvm::iplist<Statement>;
+  using iterator = StatementListType::iterator;
+  using const_iterator = StatementListType::const_iterator;
+  using reverse_iterator = StatementListType::reverse_iterator;
+  using const_reverse_iterator = StatementListType::const_reverse_iterator;
+
+  BasicBlock(const BasicBlock &) = delete;
+  BasicBlock &operator=(const BasicBlock &) = delete;
+  ~BasicBlock();
+  StatementListType &getSublist(Statement *) { return Statements(); }
+  void insertBefore(Statement *stmt, Statement *before = nullptr);
+  static BasicBlock *Create(
+      Region *parentRegion, BasicBlock *insertBefore = nullptr) {
+    return new BasicBlock(parentRegion, insertBefore);
+  }
+  const Statement *getTerminator() const;
+  Statement *getTerminator() {
+    return const_cast<Statement *>(
+        const_cast<const BasicBlock *>(this)->getTerminator());
+  }
+  void SetRegion(Region *region) { parent = region; }
+  Region *GetRegion() const { return parent; }
+  void addPred(BasicBlock *bb);
+  std::vector<BasicBlock *> &Predecessors() { return predecessors_; }
+  StatementListType &Statements() { return statementList_; }
+  BasicBlock *SplitEdge(BasicBlock *toBlock) { return nullptr; }
+
+private:
+  StatementListType statementList_;
+  std::vector<BasicBlock *> predecessors_;
+  explicit BasicBlock(Region *parentRegion, BasicBlock *insertBefore);
```

Comment:
`explicit` is only needed for (some) single-argument constructors.

Reply:
Again, coming from LLVM.  See line 76 here http://llvm.org/doxygen/BasicBlock_8h_source.html

Reply:
I don't understand.  Are you required to make this constructor `explicit` because a related one in LLVM is `explicit`?

Reply:
No. I'm saying I borrowed code from LLVM here. That's it and no implications.

Reply:
OK.

Reply:
I looked into this a bit further and found http://www.cplusplus.com/forum/general/168292/#msg843466

So prior to C++11, explicit on multi-parameter constructors didn't add anything since they couldn't be used in implicit contexts.  But C++ now allows brace initializers and with type inferencing, one can implicitly invoke a constructor with multiple parameters.  So it looks like LLVM is using `explicit` here to eliminate all but the most deliberate and text-searchable calls to the constructors of instructions.

Reply:
OK.  Should the f18 style guide be updated?  Should other f18 constructors be made `explicit`?